articles/contract-migration-guide.md

```diff
@@ -0,0 +1,239 @@
+            Account = 1,
+            /// `EntityKind::SmartContract` variant.
+            SmartContract = 2,
+        }
+        ``` 
+     - Ensure you update all key references in your contract to use this new format.
+
+2. **Entry Point Declarations:**
+    - **`EntryPointType`:**
+        - **EntryPointType::Caller:** This is the main entry point type used in Casper 2.0. It signifies that the entry point will be executed using the calling entity's context (the account that initiated the transaction)
+        - **EntryPointType::Called:** This type is used for entry points that are specifically designed to be called by other contracts. These entry points run in the context of the contract being called. This is functionally similar to the previous `Contract` type in Casper 1.x.
+        - **EntryPointType::Factory:** This is a new entry point type in Casper 2.0. It's specifically designed for contracts that create new contracts (like factory contracts). These entry points also run in the context of the called contract.
+        - The `StoredSession` entry point type is no longer available in Casper 2.0 (`EntryPointType::Caller` does not provide that functionality).
+
+    - **`EntryPointPayment`:**
+        - `EntryPointPayment::Caller`: The caller of the entry point pays for the execution cost.
+        - `EntryPointPayment::SelfOnly`: The contract itself pays for its own execution but not for any subsequent contract calls.
+        - `EntryPointPayment::SelfOnward`: The contract pays for its own execution and any subsequent contract calls it makes.
+
+    Example (CEP-18 `symbol` Entry Point):
+
+    ```rust
+    pub fn symbol() -> EntryPoint {
+        EntryPoint::new(
+            String::from(SYMBOL_ENTRY_POINT_NAME),
+            Vec::new(),
+            String::cl_type(),
+            EntryPointAccess::Public,
+            EntryPointType::Called,  // NEW
+            casper_types::EntryPointPayment::Caller, // NEW
+        )
+    }
+    ```
+   
+3. **Event Handling:**
+   - **Native Events and Message Topics:**
+     - Casper 2.0 prioritizes native event emission through message topics, offering flexibility and better integration.
+     - Utilize `runtime::emit_message` to send events on specific topics.
+   - **CES (Casper Event Standard):**
+     - CES is still supported for backward compatibility, but native events are now the recommended approach.
+
+4. **Contract Installation and Upgrade:**
+   - **`new_contract` and `add_contract_version`:**
+     - These functions now accept an additional `BTreeMap` argument for specifying message topics when creating or upgrading a contract.
+   - **Hash Types:**
+     - `ContractPackageHash` is replaced with `PackageHash`, converting to `Key::Package` instead of `Key::Hash`.
+     - `ContractHash` is replaced with `AddressableEntityHash`, corresponding to `AddressableEntity` in the global state.
+
+## Essential Upgrade Steps
+
+1. **Key Migration:**
```

Comment:
This needs to be done for existing contracts in case the contract developer/maintainer changes the way they handle keys. It is possible to write your codes in a manner that you don't need to do these migration steps.

Reply:
Added this info